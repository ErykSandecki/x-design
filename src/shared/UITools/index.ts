// components
import AlignmentArea from './AlignmentArea/AlignmentArea';
import ButtonGroup from './ButtonGroup/ButtonGroup';
import ButtonIcon from './ButtonIcon/ButtonIcon';
import Checkbox from './Checkbox/Checkbox';
import Chip from './Chip/Chip';
import Color from './Color/Color';
import ColorPicker from './ColorPicker/ColorPicker';
import FieldGroup from './FieldGroup/FieldGroup';
import GridArea from './GridArea/GridArea';
import Popover, { PopoverCompound } from './Popover/Popover';
import SectionColumn from './SectionColumn/SectionColumn';
import Section from './Section/Section';
import Select from './Select/Select';
import SelectItem from './Select/SelectItem/SelectItem';
import Tabs, { TTabsProps } from './Tabs/Tabs';
import TextField from './TextField/TextField';
import ToggleButtonGroup from './ToggleButtonGroup/ToggleButtonGroup';
import { AlignPopoverHorizontally, AlignPopoverVertically } from './Popover/enums';
import { GridColumnType } from './SectionColumn/enums';
import { TActiveCell } from './GridArea/GridAreaPopover/CellsInput/types';

export const UITools = {
  AlignPopoverHorizontally,
  AlignPopoverVertically,
  AlignmentArea,
  ButtonGroup,
  ButtonIcon,
  Checkbox,
  Chip,
  Color,
  ColorPicker,
  FieldGroup,
  GridArea,
  GridColumnType,
  Popover,
  PopoverCompound,
  Section,
  SectionColumn,
  Select,
  SelectItem,
  Tabs,
  TextField,
  ToggleButtonGroup,
};

export type TUITypes = {
  TActiveCell: TActiveCell;
  TTabsProps: TTabsProps;
};
