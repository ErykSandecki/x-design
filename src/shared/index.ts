// E2E
export * from './E2EDataAttributes/E2EDataAttribute';
export * from './E2EDataAttributes/types';
export * from './E2EDataAttributes/utils';

// Keyboards keys group
export * from './KeyboardKeysGroup/KeyboardKeysGroup';

// Scubbable input
export * from './ScrubbableInput/ScrubbableInput';

// UI
export * from './UI/components/Box/Box';
export * from './UI/components/Button/Button';
export * from './UI/components/Icon/Icon';
export * from './UI/components/Tooltip/Tooltip';
export * from './UI/components/Tooltip/core/TooltipProvider';
export * from './UI/components/Typography';
export * from './UI/components/Typography/Typography';
export * from './UI/components/Typography/enums';
export * from './UI/components/Typography/types';

// UI tools
export * from './UITools';

// Zoombox
export * from './ZoomBox/ZoomBox';
export * from './ZoomBox/utils/mousePositionRelative';
export * from './ZoomBox/utils/hasBasePosition';
export * from './ZoomBox/constants';
