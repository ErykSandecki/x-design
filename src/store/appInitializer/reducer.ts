// others
import { DEFAULT_LANGUAGE } from 'translations';
import { INIT_LANGUAGE_SUCCESS, SET_IS_APP_LOADED, SET_THEME } from './actionsType';

// types
import { TAction } from 'types';
import { TAppInitializerState, TInitLanguageSuccessAction, TSetIsAppLoadedAction, TSetThemeAction } from './types';

// utils
import { getThemePreferences } from 'utils';

const initialState: TAppInitializerState = {
  isAppLoaded: false,
  isPending: true,
  language: DEFAULT_LANGUAGE,
  theme: getThemePreferences(),
};

const initLanguageSuccess = (
  state: TAppInitializerState,
  { payload: language }: TAction<TInitLanguageSuccessAction['payload']>,
): TAppInitializerState => ({
  ...state,
  language,
});

const setIsAppLoaded = (
  state: TAppInitializerState,
  { payload: isAppLoaded }: TAction<TSetIsAppLoadedAction['payload']>,
): TAppInitializerState => ({
  ...state,
  isAppLoaded,
  isPending: false,
});

const setTheme = (
  state: TAppInitializerState,
  { payload: theme }: TAction<TSetThemeAction['payload']>,
): TAppInitializerState => ({
  ...state,
  theme,
});

const appInitializer = (state: TAppInitializerState = initialState, action: TAction): TAppInitializerState => {
  switch (action.type) {
    case INIT_LANGUAGE_SUCCESS:
      return initLanguageSuccess(state, action);
    case SET_IS_APP_LOADED:
      return setIsAppLoaded(state, action);
    case SET_THEME:
      return setTheme(state, action);
    default:
      return state;
  }
};

export default appInitializer;
