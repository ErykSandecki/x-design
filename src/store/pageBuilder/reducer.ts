// others
import {
  ADD_ELEMENT,
  APPLY_ELEMENTS_INSET_TYPE,
  APPLY_ELEMENTS_SIZE_MIN_MAX_TYPE,
  CHANGE_ALIGNMENT,
  CHANGE_BACKGROUND,
  CHANGE_LAYOUT,
  CHANGE_PARENT,
  CHANGE_POSITION,
  CLEAR_PREV_STATE,
  CHANGE_LAYOUT_BOX_SIZING,
  CHANGE_LAYOUT_GRID,
  CHANGE_INSETS,
  FIT_LAYOUT,
  FLIP_ELEMENTS,
  REDUCER_HISTORY_REDO,
  REDUCER_HISTORY_SAVE,
  REDUCER_HISTORY_UNDO,
  ROTATE_ELEMENTS,
  SELECT_ELEMENT,
  SELECT_ELEMENTS,
  SET_AREA_COORDINATES,
  RESIZE_ELEMENT,
  SET_ELEMENTS_COORDINATES,
  UNSELECT_ELEMENT,
  UPDATE_EVENTS_STATUS,
  UPDATE_PREV_STATE,
  SET_ELEMENTS_SIZES,
  APPLY_ELEMENTS_SIZE_TYPE,
  SET_ELEMENTS_SIZES_MIN_MAX,
  SET_ELEMENTS_SCORE_TO_CURRENT_SIZE,
  TOGGLE_ASPECT_RATIO,
  CHANGE_LAYOUT_ALIGNMENT,
  SET_ELEMENTS_GAP,
  APPLY_ELEMENTS_GAP_TYPE,
  CHANGE_CLIP_CONTENT,
} from './actionsType';
import { BASE_PAGE } from './constants';

// types
import { AnchorResize, AnchorRotate } from './enums';
import { KeyboardKeys, TAction } from 'types';
import {
  TAddELementAction,
  TSelectElementAction,
  TPageBuilderState,
  TUnselectElementAction,
  TSelectElementsAction,
  TSetAreaCoordinatesAction,
  TUpdateEventsStatusAction,
  TResizeElementAction,
  TRotateElementsAction,
  TChangeBackgroundAction,
  TReducerHistorySaveAction,
  TChangeAlignmentAction,
  TFlipElementsAction,
  TChangeLayoutAction,
  TSetElementsCoordinatesAction,
  TSetElementsSizesAction,
  TApplyElementsSizeTypeAction,
  TSetElementsSizesMinMaxAction,
  TSetElementsScoreToCurrentSizeAction,
  TChangeLayoutAlignmentAction,
  TSetElementsGapAction,
  TApplyElementsGapTypeAction,
  TChangeLayoutBoxSizingAction,
  TChangeLayoutGridAction,
  TChangeInsetsAction,
  TApplyElementsInsetTypeAction,
  TApplyElementsSizeMinMaxTypeAction,
  TChangeClipContentAction,
} from './types';

// utils
import { filterSelectedElements } from './utils/filterSelectedElements';
import { handleAddElement } from './utils/handleAddElement';
import { handleApplyElementsGapType } from './utils/applyElementsGapType/handleApplyElementsGapType';
import { handleApplyElementsInsetType } from './utils/applyElementsInsetType/handleApplyElementsInsetType';
import { handleApplyElementsSizeMinMaxType } from './utils/applyElementsSizeMinMaxType/handleApplyElementsSizeMinMaxType';
import { handleApplyElementsSizeType } from './utils/applyElementsSizeType/handleApplyElementsSizeType';
import { handleChangeAlignment } from './utils/changeAligment/handleChangeAlignment';
import { handleChangeBackground } from './utils/handleChangeBackground';
import { handleChangeClipContent } from './utils/handleChangeClipContent';
import { handleChangeInsets } from './utils/handleChangeInsets';
import { handleChangeLayout } from './utils/changeLayout/handleChangeLayout';
import { handleChangeLayoutAlignment } from './utils/handleChangeLayoutAlignment';
import { handleChangeLayoutBoxSizing } from './utils/handleChangeLayoutBoxSizing';
import { handleChangeLayoutGrid } from './utils/changeLayoutGrid/handleChangeLayoutGrid';
import { handleChangeParent } from './utils/changeParent/handleChangeParent';
import { handleChangePosition } from './utils/handleChangePosition';
import { handleFitLayout } from './utils/handleFitLayout';
import { handleFlipElements } from './utils/handleFlipElements';
import { handleReducerHistoryRedo } from './utils/reducerHistory/handleReducerHistoryRedo';
import { handleReducerHistorySave } from './utils/reducerHistory/handleReducerHistorySave';
import { handleReducerHistoryUndo } from './utils/reducerHistory/handleReducerHistoryUndo';
import { handleResizeElement } from './utils/resizeElement/handleResizeElement';
import { handleRotateElements } from './utils/handleRotateElements';
import { handleSetElementsCoordinates } from './utils/handleSetElementsCoordinates';
import { handleSetElementsGap } from './utils/handleSetElementsGap';
import { handleSetElementsScoreToCurrentSize } from './utils/handleSetElementsScoreToCurrentSize';
import { handleSetElementsSizes } from './utils/setElementSizes/handleSetElementsSizes';
import { handleSetElementsSizesMinMax } from './utils/handleSetElementsSizesMinMax';
import { handleToggleAspectRatio } from './utils/handleToggleAspectRatio';

const initialState: TPageBuilderState = {
  currentPage: '0',
  events: {
    canMoveElements: true,
    colorSampler: false,
    draggableElements: [],
    hoverOnElement: '-1',
    isGridDropArea: false,
    isMultipleMoving: false,
    isResizing: false,
    isRotating: false,
    possibleAnchorElementId: '-1',
    possibleAnchorPosition: null,
    possibleElement: undefined,
    possibleIndexPosition: null,
    possibleParent: null,
    pressedKey: KeyboardKeys.none,
    selectedAnchorResize: AnchorResize.none,
    selectedAnchorRotate: AnchorRotate.none,
  },
  isLoading: true,
  isPending: false,
  // @ts-ignore
  pages: {
    [BASE_PAGE.id]: {
      ...BASE_PAGE,
    },
  },
};

const addElement = (
  state: TPageBuilderState,
  { payload: element }: TAction<TAddELementAction['payload']>,
): TPageBuilderState => handleAddElement(element, state);

const applyElementsGapType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsGapTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsGapType(payload, state);

const applyElementsInsetType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsInsetTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsInsetType(payload, state);

const applyElementsSizeMinMaxType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsSizeMinMaxTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsSizeMinMaxType(payload, state);

const applyElementsSizeType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsSizeTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsSizeType(payload, state);

const changeAlignment = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeAlignmentAction['payload']>,
): TPageBuilderState => handleChangeAlignment(payload, state);

const changeBackground = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeBackgroundAction['payload']>,
): TPageBuilderState => handleChangeBackground(payload, state);

const changeClipContent = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeClipContentAction['payload']>,
): TPageBuilderState => handleChangeClipContent(payload, state);

const changeInsets = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeInsetsAction['payload']>,
): TPageBuilderState => handleChangeInsets(payload, state);

const changeLayout = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutAction['payload']>,
): TPageBuilderState => handleChangeLayout(payload, state);

const changeLayoutAlignment = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutAlignmentAction['payload']>,
): TPageBuilderState => handleChangeLayoutAlignment(payload, state);

const changeLayoutBoxSizing = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutBoxSizingAction['payload']>,
): TPageBuilderState => handleChangeLayoutBoxSizing(payload, state);

const changeLayoutGrid = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutGridAction['payload']>,
): TPageBuilderState => handleChangeLayoutGrid(payload, state);

const changeParent = (state: TPageBuilderState): TPageBuilderState => handleChangeParent(state);

const changePosition = (state: TPageBuilderState): TPageBuilderState => handleChangePosition(state);

const clearPrevState = (state: TPageBuilderState): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      prevState: undefined,
    },
  },
});

const fitLayout = (state: TPageBuilderState): TPageBuilderState => handleFitLayout(state);

const flipElements = (
  state: TPageBuilderState,
  { payload }: TAction<TFlipElementsAction['payload']>,
): TPageBuilderState => handleFlipElements(payload, state);

const reducerHistoryRedo = (state: TPageBuilderState): TPageBuilderState => handleReducerHistoryRedo(state);

const reducerHistorySave = (
  state: TPageBuilderState,
  { payload }: TAction<TReducerHistorySaveAction['payload']>,
): TPageBuilderState => handleReducerHistorySave(state, payload);

const reducerHistoryUndo = (state: TPageBuilderState): TPageBuilderState => handleReducerHistoryUndo(state);

const resizeElement = (
  state: TPageBuilderState,
  { payload: { baseCoordinates, flip, height, id, mouseCoordinates, width } }: TAction<TResizeElementAction['payload']>,
): TPageBuilderState => handleResizeElement(baseCoordinates, flip, height, width, id, mouseCoordinates, state);

const rotateElements = (
  state: TPageBuilderState,
  { payload }: TAction<TRotateElementsAction['payload']>,
): TPageBuilderState => handleRotateElements(payload, state);

const selectElement = (
  state: TPageBuilderState,
  { payload: selectedElement }: TAction<TSelectElementAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: filterSelectedElements(
        [...state.pages[state.currentPage].selectedElements, selectedElement],
        state,
      ),
    },
  },
});

const selectElements = (
  state: TPageBuilderState,
  { payload: selectedElements }: TAction<TSelectElementsAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: filterSelectedElements(selectedElements, state),
    },
  },
});

const setAreaCoordinates = (
  state: TPageBuilderState,
  { payload: areaCoordinates }: TAction<TSetAreaCoordinatesAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      areaCoordinates: {
        ...state.pages[state.currentPage].areaCoordinates,
        ...areaCoordinates,
      },
    },
  },
});

const setElementsCoordinates = (
  state: TPageBuilderState,
  { payload: coordinates }: TAction<TSetElementsCoordinatesAction['payload']>,
): TPageBuilderState => handleSetElementsCoordinates(coordinates, state);

const setElementsGap = (
  state: TPageBuilderState,
  { payload }: TAction<TSetElementsGapAction['payload']>,
): TPageBuilderState => handleSetElementsGap(payload, state);

const setElementsScoreToCurrentSize = (
  state: TPageBuilderState,
  { payload: { scoreType, sizeType } }: TAction<TSetElementsScoreToCurrentSizeAction['payload']>,
): TPageBuilderState => handleSetElementsScoreToCurrentSize(scoreType, sizeType, state);

const setElementsSizes = (
  state: TPageBuilderState,
  { payload: { sizeType, value } }: TAction<TSetElementsSizesAction['payload']>,
): TPageBuilderState => handleSetElementsSizes(sizeType, state, value);

const setElementsSizesMinMax = (
  state: TPageBuilderState,
  { payload: { scoreType, sizeType, value } }: TAction<TSetElementsSizesMinMaxAction['payload']>,
): TPageBuilderState => handleSetElementsSizesMinMax(scoreType, sizeType, state, value);

const toggleAspectRatio = (state: TPageBuilderState): TPageBuilderState => handleToggleAspectRatio(state);

const updateEventsStatus = (
  state: TPageBuilderState,
  { payload: events }: TAction<TUpdateEventsStatusAction['payload']>,
): TPageBuilderState => ({
  ...state,
  events: {
    ...state.events,
    ...events,
  },
});

const updatePrevState = (state: TPageBuilderState): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      prevState: state.pages[state.currentPage],
    },
  },
});

const unselectElement = (
  state: TPageBuilderState,
  { payload: id }: TAction<TUnselectElementAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: state.pages[state.currentPage].selectedElements.filter((element) => element.id !== id),
    },
  },
});

const pageBuilder = (state: TPageBuilderState = initialState, action: TAction): TPageBuilderState => {
  switch (action.type) {
    case ADD_ELEMENT:
      return addElement(state, action);
    case APPLY_ELEMENTS_GAP_TYPE:
      return applyElementsGapType(state, action);
    case APPLY_ELEMENTS_INSET_TYPE:
      return applyElementsInsetType(state, action);
    case APPLY_ELEMENTS_SIZE_MIN_MAX_TYPE:
      return applyElementsSizeMinMaxType(state, action);
    case APPLY_ELEMENTS_SIZE_TYPE:
      return applyElementsSizeType(state, action);
    case CHANGE_ALIGNMENT:
      return changeAlignment(state, action);
    case CHANGE_BACKGROUND:
      return changeBackground(state, action);
    case CHANGE_CLIP_CONTENT:
      return changeClipContent(state, action);
    case CHANGE_INSETS:
      return changeInsets(state, action);
    case CHANGE_LAYOUT:
      return changeLayout(state, action);
    case CHANGE_LAYOUT_ALIGNMENT:
      return changeLayoutAlignment(state, action);
    case CHANGE_LAYOUT_BOX_SIZING:
      return changeLayoutBoxSizing(state, action);
    case CHANGE_LAYOUT_GRID:
      return changeLayoutGrid(state, action);
    case CHANGE_PARENT:
      return changeParent(state);
    case CHANGE_POSITION:
      return changePosition(state);
    case CLEAR_PREV_STATE:
      return clearPrevState(state);
    case FIT_LAYOUT:
      return fitLayout(state);
    case FLIP_ELEMENTS:
      return flipElements(state, action);
    case REDUCER_HISTORY_REDO:
      return reducerHistoryRedo(state);
    case REDUCER_HISTORY_SAVE:
      return reducerHistorySave(state, action);
    case REDUCER_HISTORY_UNDO:
      return reducerHistoryUndo(state);
    case RESIZE_ELEMENT:
      return resizeElement(state, action);
    case ROTATE_ELEMENTS:
      return rotateElements(state, action);
    case SELECT_ELEMENT:
      return selectElement(state, action);
    case SELECT_ELEMENTS:
      return selectElements(state, action);
    case SET_AREA_COORDINATES:
      return setAreaCoordinates(state, action);
    case SET_ELEMENTS_COORDINATES:
      return setElementsCoordinates(state, action);
    case SET_ELEMENTS_GAP:
      return setElementsGap(state, action);
    case SET_ELEMENTS_SCORE_TO_CURRENT_SIZE:
      return setElementsScoreToCurrentSize(state, action);
    case SET_ELEMENTS_SIZES:
      return setElementsSizes(state, action);
    case SET_ELEMENTS_SIZES_MIN_MAX:
      return setElementsSizesMinMax(state, action);
    case TOGGLE_ASPECT_RATIO:
      return toggleAspectRatio(state);
    case UPDATE_EVENTS_STATUS:
      return updateEventsStatus(state, action);
    case UPDATE_PREV_STATE:
      return updatePrevState(state);
    case UNSELECT_ELEMENT:
      return unselectElement(state, action);
    default:
      return state;
  }
};

export default pageBuilder;
