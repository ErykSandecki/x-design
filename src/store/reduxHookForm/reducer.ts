import { mapValues, omit } from 'lodash';

// others
import {
  BLUR,
  CHANGE,
  CLEAR_FIELDS,
  DESTROY_FORM,
  FOCUS,
  INIT_FIELD,
  MOUNT_FORM,
  SET_PENDING,
  SET_PENDING_FIELD,
  SET_TOUCHED_FIELD,
  SUBMIT,
  SUBMIT_ERROR,
  SUBMIT_SUCCESS,
  UPDATE_ASYNC_ERRORS,
  UPDATE_FORM_VALIDATOR,
  UPDATE_SYNC_ERRORS,
} from './actionsType';

// types
import { TAction } from 'types/redux';
import {
  TClearFieldsAction,
  TDestroyFormAction,
  TInitFieldAction,
  TMountFormAction,
  TReduxHookFormState,
  TSubmitAction,
  TSubmitErrorAction,
  TSubmitSuccessAction,
  TUpdateFieldAction,
  TUpdateFormAction,
} from './types';

// utils
import { getUpdatedFieldsState, getFieldsWithModifiedAttributes, notifyFields } from './utils';

const initialState: TReduxHookFormState = {};

const clearFields = (
  state: TReduxHookFormState,
  { payload }: TAction<TClearFieldsAction['payload']>,
): TReduxHookFormState => {
  const names = payload!.names;
  const formName = payload!.formName;

  const fields = mapValues(state[payload!.formName].fields, (field, name) =>
    names.includes(name)
      ? {
          ...field,
          asyncErrors: [],
          previousValue: field.emptyValue,
          syncErrors: [],
          touched: false,
          value: field.emptyValue,
        }
      : field,
  );

  return {
    ...state,
    [formName]: {
      ...state[formName],
      fields,
    },
  };
};

const destroyForm = (
  state: TReduxHookFormState,
  { payload: formName }: TAction<TDestroyFormAction['payload']>,
): TReduxHookFormState => {
  const newState = omit({ ...state }, [formName]);

  return {
    ...newState,
  };
};

const initField = (
  state: TReduxHookFormState,
  { payload: { formName, field, name } }: TAction<TInitFieldAction['payload']>,
): TReduxHookFormState => {
  if (!state[formName]) {
    return state;
  }

  return {
    ...state,
    [formName]: {
      ...state[formName],
      fields: getUpdatedFieldsState(field, formName, name, state),
    },
  };
};

const mountForm = (
  state: TReduxHookFormState,
  { payload }: TAction<TMountFormAction['payload']>,
): TReduxHookFormState => ({ ...state, ...payload });

const submit = (
  state: TReduxHookFormState,
  { payload: formName }: TAction<TSubmitAction['payload']>,
): TReduxHookFormState => {
  notifyFields(formName, state, 'before');

  return {
    ...state,
    [formName]: {
      ...state[formName],
      error: '',
      fields: getFieldsWithModifiedAttributes(formName, state),
      isPending: true,
    },
  };
};

const submitError = (
  state: TReduxHookFormState,
  { payload: { error, formName } }: TAction<TSubmitErrorAction['payload']>,
): TReduxHookFormState => {
  notifyFields(formName, state, 'after');
  return {
    ...state,
    [formName]: {
      ...state[formName],
      error,
      isPending: false,
    },
  };
};

const submitSuccess = (
  state: TReduxHookFormState,
  { payload: formName }: TAction<TSubmitSuccessAction['payload']>,
): TReduxHookFormState => {
  notifyFields(formName, state, 'after');
  return {
    ...state,
    [formName]: {
      ...state[formName],
      isPending: false,
    },
  };
};

const updateField = (
  state: TReduxHookFormState,
  { payload: { formName, field, name } }: TAction<TUpdateFieldAction['payload']>,
): TReduxHookFormState => ({
  ...state,
  [formName]: {
    ...state[formName],
    fields: getUpdatedFieldsState(field, formName, name, state),
  },
});

const updateForm = (
  state: TReduxHookFormState,
  { payload: { form, formName } }: TAction<TUpdateFormAction['payload']>,
): TReduxHookFormState => ({
  ...state,
  [formName]: {
    ...state[formName],
    ...form,
  },
});

export const reduxHookForm = (state: TReduxHookFormState = initialState, action: TAction): TReduxHookFormState => {
  switch (action.type) {
    case CLEAR_FIELDS:
      return clearFields(state, action);
    case DESTROY_FORM:
      return destroyForm(state, action);
    case INIT_FIELD:
      return initField(state, action);
    case MOUNT_FORM:
      return mountForm(state, action);
    case SUBMIT:
      return submit(state, action);
    case SUBMIT_ERROR:
      return submitError(state, action);
    case SUBMIT_SUCCESS:
      return submitSuccess(state, action);
    case BLUR:
    case CHANGE:
    case FOCUS:
    case SET_PENDING_FIELD:
    case SET_TOUCHED_FIELD:
    case UPDATE_ASYNC_ERRORS:
    case UPDATE_SYNC_ERRORS:
      return updateField(state, action);
    case SET_PENDING:
    case UPDATE_FORM_VALIDATOR:
      return updateForm(state, action);
    default:
      return state;
  }
};

export default reduxHookForm;
