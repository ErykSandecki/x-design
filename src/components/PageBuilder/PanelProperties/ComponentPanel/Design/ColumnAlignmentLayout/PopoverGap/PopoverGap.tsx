import { FC, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import { useTranslation } from 'react-i18next';

// components
import { UITools } from 'shared';

// others
import { translationNameSpace } from './constants';

// store
import { applyElementsGapType } from 'store/pageBuilder/actions';

// types
import { PopoverItem } from '../enums';
import { TGap, TGapProperties } from 'types';

// utils
import { isPureNumber } from 'utils';

const { PopoverCompound } = UITools;

export type TPopoverGapProps = {
  gap: keyof TGap;
  gapProperties: TGapProperties;
  isMixed: boolean;
};

const PopoverGap: FC<TPopoverGapProps> = ({ gap, gapProperties, isMixed }) => {
  const icon = useMemo(() => (gap === 'column' ? 'GapColumns' : 'GapRows'), []);
  const dispatch = useDispatch();
  const { t } = useTranslation();

  return (
    <>
      <PopoverCompound.PopoverItem
        icon={icon}
        index={PopoverItem.fixed}
        onClick={() => dispatch(applyElementsGapType(gap, 'fixed'))}
        selected={!isMixed && isPureNumber(gapProperties.value)}
        text={t(`${translationNameSpace}.${gap}.1`, { value: gapProperties.value })}
      />
    </>
  );
};

export default PopoverGap;
