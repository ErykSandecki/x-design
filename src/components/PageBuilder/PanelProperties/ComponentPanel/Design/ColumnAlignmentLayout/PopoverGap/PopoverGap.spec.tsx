import { noop } from 'lodash';

// components
import PopoverGap from './PopoverGap';
import { UITools } from 'shared';

// mocks
import { gapMock } from 'test/mocks/reducer/pageBuilderMock';

// utils
import { customRender } from 'test';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

describe('PopoverGap snapshots', () => {
  it('should render PopoverGap', () => {
    // before
    const { asFragment } = customRender(
      <UITools.PopoverCompound.PopoverRoot setSelected={noop}>
        <PopoverGap gap="column" gapProperties={gapMock} isMixed={false} />
      </UITools.PopoverCompound.PopoverRoot>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render wiht mixed value', () => {
    // before
    const { asFragment } = customRender(
      <UITools.PopoverCompound.PopoverRoot setSelected={noop}>
        <PopoverGap gap="column" gapProperties={gapMock} isMixed />
      </UITools.PopoverCompound.PopoverRoot>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
