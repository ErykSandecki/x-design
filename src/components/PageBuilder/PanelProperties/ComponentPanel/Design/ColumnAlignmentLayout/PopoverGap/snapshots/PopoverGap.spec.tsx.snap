// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverGap snapshots should render PopoverGap 1`] = `
<DocumentFragment>
  <div
    class="css-1wfopar PopoverItem PopoverItem--light"
    data-test-box=""
    data-test-popover-item="0"
  >
    <div
      class="css-1jnxvbp PopoverItem__checkIcon PopoverItem__checkIcon--light PopoverItem__checkIcon--selected PopoverItem__checkIcon--selected--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
      data-test-icon="check"
      height="12"
      name="Check"
      width="12"
    />
    <div
      class="css-1jnxvbp PopoverItem__icon PopoverItem__icon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
      data-test-icon="gap-columns"
      height="12"
      name="GapColumns"
      width="12"
    />
    <small
      class="css-1jnxvbp PopoverItem__text PopoverItem__text--light Typography Typography--light Typography--neutral1 Typography--neutral1--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--small Typography--small--light"
      data-test-text=""
      style="text-align: inherit;"
    >
      Fixed column gap (0)
    </small>
  </div>
</DocumentFragment>
`;

exports[`PopoverGap snapshots should render wiht mixed value 1`] = `
<DocumentFragment>
  <div
    class="css-1wfopar PopoverItem PopoverItem--light"
    data-test-box=""
    data-test-popover-item="0"
  >
    <div
      class="css-1jnxvbp PopoverItem__checkIcon PopoverItem__checkIcon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
      data-test-icon="check"
      height="12"
      name="Check"
      width="12"
    />
    <div
      class="css-1jnxvbp PopoverItem__icon PopoverItem__icon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
      data-test-icon="gap-columns"
      height="12"
      name="GapColumns"
      width="12"
    />
    <small
      class="css-1jnxvbp PopoverItem__text PopoverItem__text--light Typography Typography--light Typography--neutral1 Typography--neutral1--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--small Typography--small--light"
      data-test-text=""
      style="text-align: inherit;"
    >
      Fixed column gap (0)
    </small>
  </div>
</DocumentFragment>
`;
